### PR TITLE
Export generateServiceProviderMetadata

### PR DESCRIPTION
**Pull request by nwalters512**
# Description

This PR makes the change proposed in #336 and directly exports `generateServiceProviderMetadata`. This is useful for generating metadata without constructing a complete and valid `SAML` instance.

For this function to be maximally useful on its own, I made some changes so that it would behave the same as if it were used on an instance of `SAML`. That is, for some set of options, `new SAML(options).generateServiceProviderMetadata() === generateServiceProviderMetadata(options)` should be true. This mostly entailed making additional options optional and providing defaults that matched those set in `SAML#initialize()`.

# Checklist:

- Issue Addressed: #336
- Link to SAML spec: N/A
- Tests included? [x]
- Documentation updated? [x]

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -84,6 +84,7 @@ export type CertCallback = (
 export interface MandatorySamlOptions {
   cert: string | string[] | CertCallback;
   issuer: string;
+  callbackUrl: string;
```
Comment (nwalters512) [inline src/types.ts:87]: I noticed while writing the test that `callackUrl` was incorrectly marked as optional given that there's a runtime check to make sure it's provided:

https://github.com/node-saml/node-saml/blob/e691ccf39b9f3521268fc31b8b8f099beae4654c/src/saml.ts#L96

I acknowledge that this isn't under the original scope of the PR, I'd be happy to move it to another one if requested.

File: src/metadata.ts [hunk 2]
```diff
@@ -15,13 +16,14 @@ export const generateServiceProviderMetadata = (
...
     logoutCallbackUrl,
-    identifierFormat,
-    wantAssertionsSigned,
     decryptionPvk,
     privateKey,
     metadataContactPerson,
     metadataOrganization,
-    generateUniqueId,
+    // The below defaults are selected to match the defaults in the `SAML` class.
+    identifierFormat = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress",
```
Comment (cjbarth) [inline src/metadata.ts:24]: I'd prefer to keep all the defaults in a single location. If they need to be duplicated here, then perhaps we should put a run-time check to make sure they are populated and leave the onus on the one calling this method directly to know what their doing.

Comment (nwalters512) [reply src/metadata.ts:24]: Is there any particular way you'd like me to share these defaults? Maybe declare them in a constant somewhere? Is there precedent for this?

I'd rather not leave this to the caller; if I'm not explicitly configuring `identifierFormat` for instances of `SAML`, I wouldn't expect to have to configure them here either.

Comment (cjbarth) [reply src/metadata.ts:24]: I think I might be OK with a set of constants in a new file: `constants.ts`. I see we've already polluted the `types.ts` file a little with a class that should probably be in `utilities.ts`, but I don't want that to be a pattern. It also doesn't seem to be a great fit for `utilities.ts`. I'm interested in your thoughts.

I really want to make sure that the next guy (or future me) doesn't think there is just one place for defaults and forgets to change something.

Comment (nwalters512) [reply src/metadata.ts:24]: I think your motivation is spot on. I think a `constants.ts` file with exports like `DEFAULT_IDENTIFIER_FORMAT` and `DEFAULT_WANT_ASSERTIONS_SIGNED` would be a good idea; I'll make that change if you're good with it!

Comment (nwalters512) [reply src/metadata.ts]: @cjbarth done in commit d38a77d.

**Comment by nwalters512** [comment]
Looks like some tests were failing in CI. However, they also fail on `master` both locally and on CI, so I'm guessing that isn't anything to do with my changes.

**Comment by cjbarth** [comment]
@nwalters512 , I think everything checks out now. Can you give this a final once-over to make sure that the merge I did still looks good? I want to do a release and this code is close to landing, so I want to include it.

**Comment by nwalters512** [comment]
@cjbarth thanks for getting this up to date with master. This looks ready to go to me!

**Pull request merged**